crates/core/src/db/relational_db.rs

```diff
@@ -95,33 +95,56 @@ impl RelationalDB {
             let mut last_commit_offset = None;
             let mut last_hash: Option<Hash> = None;
             if let Some(message_log) = &message_log {
-                let message_log = message_log.lock().unwrap();
+                let mut message_log = message_log.lock().unwrap();
                 let max_offset = message_log.open_segment_max_offset;
-                for commit in commit_log::Iter::from(message_log.segments()) {
-                    let commit = commit?;
-
-                    segment_index += 1;
-                    last_hash = commit.parent_commit_hash;
-                    last_commit_offset = Some(commit.commit_offset);
-                    for transaction in commit.transactions {
-                        transaction_offset += 1;
-                        // NOTE: Although I am creating a blobstore transaction in a
-                        // one to one fashion for each message log transaction, this
-                        // is just to reduce memory usage while inserting. We don't
-                        // really care about inserting these transactionally as long
-                        // as all of the writes get inserted.
-                        datastore.replay_transaction(&transaction, odb.clone())?;
-
-                        let percentage = f64::floor((segment_index as f64 / max_offset as f64) * 100.0) as i32;
-                        if percentage > last_logged_percentage && percentage % 10 == 0 {
-                            last_logged_percentage = percentage;
-                            log::debug!(
-                                "[{}] Loaded {}% ({}/{})",
-                                address,
-                                percentage,
-                                transaction_offset,
-                                max_offset
-                            );
+                let total_segments = message_log.total_segments();
+
+                'replay: for (segment_offset, segment) in message_log.segments().enumerate() {
+                    for commit in segment.try_into_iter_commits()? {
+                        let commit = match commit {
+                            Ok(commit) => commit,
+                            Err(e) if e.kind() == io::ErrorKind::InvalidData => {
```

Comment:
I did consider this, but it turns out we need to handle more cases here:

1. The `Commit` envelope decodes successfully, but is actually invalid
2. The `Commit` envelope fails to decode
3. The file is EOF _before_ commit decoding even starts

It's not obvious how to unify this into a single error type (at least not without some major refactoring), and I'm not even sure it would be of value -- we would not gain any exhaustiveness checking.

I hope to have improved the upstream error reporting (i.e. what comes out of `RelationalDB::open`) in the recent commits, though.

Reply:
This line of thought ok with you Joshua?

Reply:
I have formulated this differently in #554, which is a large enough diff to submit separately:

https://github.com/clockworklabs/SpacetimeDB/pull/554/files#diff-37c945f95a02769137efa2981278629b8e8aa25d24504f0c4210624b552dda93R473

This keeps enough state to keep handling of the different cases close to each other. It may also become general iteration at some point, but I’m not sure.